src/store/SwapMetaView.cc

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "base/Raw.h"
+#include "base/TextException.h"
+#include "debug/Stream.h"
+#include "sbuf/Stream.h"
+#include "store/SwapMetaView.h"
+
+#include <iostream>
+
+namespace Store {
+
+/// properly reports or rejects a problematic raw swap meta field type
+static void
+HandleBadRawType(const RawSwapMetaType type)
+{
+    if (ReservedSwapMetaType(type)) {
+        debugs(20, 3, "ignoring swap meta field with a reserved type: " << int(type));
+        return;
+    }
+
+    if (DeprecatedSwapMetaType(type)) {
+        debugs(20, DBG_CRITICAL, "ERROR: Ignoring swap meta field with a deprecated type: " << int(type));
+        return;
+    }
+
+    // TODO: Instead of assuming that all future swap meta types can be ignored
+    // (and some should be reported at level-0/1), future-proof this code by
+    // using a type bit to define whether to silently ignore a swap meta field
+    // with that type (or even the whole Store entry with that field).
+
+    const auto typeValuesWeMightAdd = 10;
+    // compute "type > RawSwapMetaTypeTop() + typeValuesWeMightAdd" w/o overflow
+    if (type >= typeValuesWeMightAdd && type - typeValuesWeMightAdd > RawSwapMetaTypeTop()) {
+        debugs(20, DBG_CRITICAL, "ERROR: Malformed cache storage; ignoring swap meta field with an unexpected type: " << int(type));
+        return;
+    }
+
+    if (type > RawSwapMetaTypeTop()) {
+        debugs(20, 3, "ignoring swap meta field with a presumed future type: " << int(type));
+        return;
+    }
+
+    Assure(type <= RawSwapMetaTypeBottom);
+    debugs(20, DBG_CRITICAL, "ERROR: Malformed cache storage; ignoring swap meta field with an invalid type: " << int(type));
+}
+
+} // namespace Store
+
+Store::SwapMetaView::SwapMetaView(const void * const begin, const void * const end)
+{
+    auto input = static_cast<const char *>(begin);
+
+    SwapMetaExtract(rawType, input, end);
+    if (HonoredSwapMetaType(rawType))
+        type = static_cast<SwapMetaType>(rawType);
+    else
+        HandleBadRawType(rawType); // and leave type as STORE_META_VOID
+
+    RawSwapMetaLength lengthOrGarbage = 0;
+    SwapMetaExtract(lengthOrGarbage, input, end);
+    if (lengthOrGarbage < 0)
+        throw TextException("negative swap meta field length value", Here());
+    if (uint64_t(lengthOrGarbage) > SwapMetaFieldValueLengthMax)
+        throw TextException("huge swap meta field length value", Here());
+    if (input + lengthOrGarbage > end)
+        throw TextException("truncated swap meta field value", Here());
+    rawLength = static_cast<size_t>(lengthOrGarbage);
+
+    Assure(input > begin);
+    Assure(input + rawLength <= end);
```

Comment:
> This is an input parser performing validation. Assure is not supposed to be used for validation.

These Assure() statements are not about input validation: If at least one of those conditions is false, we would need to fix Squid _code_. Once we get to this Assure() lines/code, we state that the previous input validation left us a usable/valid TLV value pointer. We should use assert() or Assure() for that code logic check, of course.

In commit 06a249b, I added explicit condition descriptions in hope to make progress, but I believe the code was sufficiently self-documented without them. IMO, the level of effort required to figure out that the asserted conditions are implied by the above code is acceptable:

```C++
auto input = static_cast<const char *>(begin); // no const implies that we are changing this variable
SwapMetaExtract(... input ...); // we are probably changing it here (can be confirmed via documentation)
SwapMetaExtract(... input ...); // yes, definitely changing or we would not extract the second field here!
Assure(input > begin); // aha, we made progress but (together with the other assertion) stayed in the range
```

```C++
if (input + lengthOrGarbage > end) // check whether the value component ends before (or where) the buffer does
    throw ...
rawLength = lengthOrGarbage; // OK, validated and copied the length field, committing to using it
Assure(input + rawLength <= end); // aha, (together with the other assertion) we stayed in the range
```

Reply:
> > This is an input parser performing validation. Assure is not supposed to be used for validation.
> 
> These Assure() statements are not about input validation: If at least one of those conditions is false, we would need to fix Squid _code_.

Wrong. These statements are performing conditional tests on input variables still containing the unaltered raw input bits (rawLength, lengthOrGarbage). That is by definition input validation - regardless of whether any previous validation checks have been done to the same variable.

> Once we get to this Assure() lines/code, we state that the previous input validation left us a usable/valid TLV value pointer. We should use assert() or Assure() for that code logic check, of course.

If you are so sure the values are impossible, use assert().

If the values *can* occur, you have defined them as bad input. This *function* is about input validation - using a calculation involving *raw* input bits .... use of Assure() is prohibited by its definition.

Reply:
> use assert()

Done in 23b01064bb5cbd6e9b7911373385f2b127d56a0b in hope to make progress.



> These statements are performing conditional tests on input variables still containing the unaltered raw input bits (rawLength, lengthOrGarbage). 

Correct.

> That is by definition input validation

This is not how the technical term "input validation" is [defined](https://www.whitehatsec.com/glossary/content/input-validation). Input validation checks input. This code does not check input. This code checks code logic. Whether those code logic checks are using variables containing input is immaterial (and nearly all checks will naturally contain variables containing input or its derivatives). Here is a simple example to illustrate the difference (the `//` comments state the obvious and would not exist in real code):

```C++
const auto input = readInput();
  
// This is input validation.
if (badInput(input))
    throw "bad input";

/* we are dealing with good input now */

// This is code logic validation:
// We assert that good input is positive.
// If we are wrong, the assertion will fail, and we will need to fix the bug.
assert(input > 0);

const auto size = input - 1;
```

This is basic/standard development terminology and practice. We should not be arguing about it!

> If you are so sure the values are impossible, use assert().

The structure of the above request mismatches the difference between assert() and Assure(). Here is the official description of Assure() explaining the true difference between those two assertions: Assure() is "Like assert() but throws an exception instead of aborting the process. Use this macro to detect code logic mistakes (i.e. bugs) where aborting the current AsyncJob or a similar task is unlikely to jeopardize Squid service integrity."

The choice between assert() and Assure() is never about being sure that the condition is true -- both macros are placed in the code when the developer believes the condition must be true.